File: taskrs-server/src/models/request_filter.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RequestFilter<T> {
    pub query: Option<String>,
    pub order_by: Option<T>,
    pub order: Option<Order>,
    pub page: Option<i64>,
    pub limit: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Order {
    Ascending,
    Descending,
}

File: taskrs-web/src/app/core/models/request-filter.ts
import {HttpParams} from '@angular/common/http';

export interface RequestFilter {
    query: string | null,
    orderBy: string | null,
    order: 'ascending' | 'descending' | null,
    page: number | null,
    limit: number | null,
}

export function createRequestFilterParams(filter: RequestFilter | null): HttpParams {
    let params = new HttpParams();

    if (filter === null) {
        return params;
    }

    if (filter.query !== null) {
        params = params.set('query', filter.query);
    }
    if (filter.orderBy !== null) {
        params = params.set('orderBy', filter.orderBy);
    }
    if (filter.page !== null) {
        params = params.set('page', filter.page);
    }
    if (filter.page !== null) {
        params = params.set('page', filter.page);
    }
    if (filter.limit !== null) {
        params = params.set('limit', filter.limit);
    }

    return params;
}
